README.md
zot_bib_web: Interactive web bibliographies with Zotero
============================================

Example Installations
----------------------------------------

[Lab website at Penn State](http://acs.ist.psu.edu/wp/pub/)

[Personal website](http://david-reitter.com/pub/)


Documentation
----------------------------------------

http://zot-bib-web.readthedocs.io/


Features
----------------------------------------

This tool generates interactive web bibliographies based on one or
more collections in a Zotero repository.

Collections can be maintained by groups of people, using Zotero's web
interface or their desktop applications.

Zot_bib_web does not depend on any third-party web server.  The
generated bibliographies load quickly because they are stored as static
files along with the rest of your website.

This makes a good source for webcrawlers, including Google Scholar and CiteSeer.

Setup is very easy for anyone who runs their own website and knows
how to use a command line (shell).

The easiest way to use it is to call zot.py with the key of a public
Zotero collection.  It will make a zotero-bib.html file.  Copy this,
along with the "site" and "files" (if any) directory to your webserver.

Bibliographies
-  can be ordered by collection, by publication year, or by
   publication type (e.g., journal articles first),
-  are editable in Zotero by one or more users,
-  are interactively searchable,
-  can be linked to PDF documents or other URLs,
-  have records for BibTex, EndNote and Wikipedia, and
-  can be exported to HTML or pushed to a Wordpress database.

The following formats and info can also be included:
BibTex, RIS (for EndNote etc.), Wikipedia markup, abstract, PDF,
APA and MLA format.

A readable bibliographic entry (e.g., in APA style) is always shown.
COInS is always included (invisible).

The content generated is static.  This program is meant to
be run periodically.

Support for other languages is provided; by default for German and English.

A tool to push the resulting bibliography to a Wordpress installation
is also provided.

zot_bib_web is available on Github.


Demo
----------------------------------------

- View the HTML files in the demo folder for some examples of
  bibliographies.  Their respective settings files and CSS style files
  are included.

- Run

  ./zot.py --settings demo/settings3.py

  to see it in action.


License and Donations
----------------------------------------

- Use and modify this software free of charge.
- No warranty is provided whatsoever.
- Please e-mail david.reitter@gmail.com a link to the bibliography on
  your website if you decide to use zot_bib_web.

- You may use this software for free.  A donation is suggested, either
  via Bitcoin (1CsboLGieSnxWeVC4dFZjBGPQEn5Uyfsia) or with a
  credit card at http://aquamacs.org/donate


Requirements
----------------------------------------
- Python 2.7 or 3.6+
- Pyzotero.
To install Pyzotero, a library for python:

		sudo pip install pyzotero

	or:

		sudo easy_install pyzotero

- A Zotero collection with your bibliography (as user or as group)

Optional:
- dateutils package for Python (improves date parsing if present)


Setup
-----------------------------------------
- ensure zot.py is executable (chmod ug+x zot.py)

- Try it out.  From a unix-like command-line, do this:

		./zot.py --group 160464 DTDTV2EP

   Then view zotero-bib.html in a browser.
   If that looks good, move on to the next steps for configuration.

- In a new file called settings.py, add configuration as documented in
  the file settings_example.py.
  Go to zotero.org to get your API secret key and your user or library
  IDs.  It's easy: see the top of settings.example.py for details.
  If settings.py is set up, you can call zot.py without arguments.

  Alternatively, you can use give the primary settings in arguments to
  the program.



Bibliography in Zotero
-----------------------------------------
- With Zotero, create a bibliography and note its ID (e.g., from the
  URL in the Zotero web interface).  Example: `MGID90AT`.
  This ID is what you need for the "toplevelfilter" variable in
  settings.py.

- You can add sub-collections to your bibliography.

- If you format ordered by collections, giving them an order may be
  helpful.  You can name collections starting with
  a number: "10 Social Psychology".

Here's an example of a bibliography structure:

	My Publications [MGID90AT]
		10 Selected Works
		15 In Preparation / Under Review
		20 Refereed Works by Topic
			Semantics
			Parsing
			Dialogue
			Machine Learning
		30 Theses
		40 Talks (Without Paper)

To see this, use the provided settings.py as an example.



Configuration options
-----------------------------------------

- Configuration takes place in a settings file, by default named
  settings.py.

- Call ./zot.py --help to see a list of command-line options.

- Please see settings.example.py for documentation.  A few options are
  discussed in the following.

- You can order our bibliography by sub-collection, by year, or by
  publication type (e.g., journal articles first, then conference
  papers). Even within the higher-level categories you can sort your
  bibliographic entries as you wish.  Use the "sort_criteria" and
  "show_top_section_headings" settings.

- You can choose a different formatting convention.  Default is APA format.

- At the top, a search box and a set of shortcuts may be displayed:
  for example, several years or spans of years, publication types, or
  subcollections that indicate topic areas.  A visitor may click on
  these to quickly filter the bibliography.

- Configure the shortcuts shown at the top using the show_shortcuts
   variable.  'collection', 'type', 'year', 'venue', 'venue_short', and
   'tags' are supported values, in addition to more fine-grained lists
   of values you can create using the shortcut() function.  You can
   give the list of values or ranges (for years), their order, and
   some filtering to only show the most common ones.  See
   settings.example.py for a detailed example.

- There are several more options.  Again, see settings.example.py.




Deployment to a web site
-----------------------------------------

- Upload the site folder or its contents to a public place on your web server.
  By default, /site/... is the assumed URL.

To generate HTML and include it in a website:

-  run zot.py once/on demand, or install as cron job or service on a server
Do not run it more than once a day.  Configure it directly in zot.py,
or in a separate file settings.py to make upgrading simple.

- include the resulting file zotero-bib.html (or as configured) in
  your website as you see fit.  You may also include individual
  collection files, which are also generated.   You can configure
  zot.py to generate a complete HTML document, or just a portion of
  it.  Zot_bib_web generates HTML5 content.

- Style your bibliography using CSS.  An example style
  file is included (see site/ directory).



Wordpress Support
-----------------------------------------

This package can push directly to a Wordpress site.  A separate
program "push.py" is included to do this.

Follow these steps:

1.  Set up zot.py to generate a bibliography you like.
	Call zot.py --full to generate a complete zotero-bib.html file
	for debugging purposes.  Configure settings.py to not generate
	the full HTML code.
2. Install the wpautop-control plugin (or a similar plugin) to make
	sure that WP will not insert paragraph breaks at various places
	in the bibliography.  With this plugin, you will need to add a
	"custom field" to the page created in the  next step (Choose
	"Screen Options" at the top of the page view, enable custom
	fields.  Then find custom fields at the very bottom of the page
	and add a "wpautop" field with value "no".
3.  Create a WP page or a post for the bibliography. Insert
	[zot_bib_web COLLECTION] where you'd like the bibliography
	inserted.  Replace COLLECTION with the ID of the collection.
	(More options: see push.py)
4. Copy the style sheet contents (in site/) to your Wordpress theme
	(select "editor", or "Additional CSS").
5. Configure settings.py so that jquery and other files are available
	on the web server.  Typically, this would be
		jquery_path = "../wp-includes/js/jquery/jquery.js"
	clipboard.js and clippy.svg: You may refer to a public URL or serve
	the files yourself.
6.  Configure push.py (at the top).  You will need to know a few simple
	details about your WP installation.
7.  Run push.py regularly or on demand.  It will call zot.py
	automatically and then update the page in WP.





How it works
-----------------------------------------

Zot.py produces annotated, beautiful bibliographies for the web from a
Zotero collection.  It is designed for individuals and for research
groups.

This program will load settings.py for its configuration.
It will retrieve all publications for the given user or
group.  If subcollections are present, each subcollection will form a
separate section.  The top-level collection can be ignored (depending
on configuration.)

The output of zot.py consists of HTML: either a full document, or a snippet, as
configured.

Each bibliographic entry is annotated with a bib record and an abstract,
each of which can be revealed by the reader (client-side javascript).



Author
-----------------------------------------
David Reitter, Penn State


ChangeLog.md
Changes to Zot_bib_web

https://github.com/davidswelt/zot_bib_web

Author/Contact:
	Dr. David Reitter, College of Information Sciences & Technology,
	Penn State
	reitter@psu.edu


zot_bib_web 2.1.0
-----------------------------------------

- Add --quiet and --verbose arguments and a `verbosity' configuration
  variable.  Output is now primarily made to stderr.  Other improvements
  to progress output.

- Links in bib entries are now clickable.

- Fix bugs in connection with empty author and title fields.

- Fix problems with Windows in connection with non-ASCII characters

- Various bugfixes.

-- David Reitter <reitter@psu.edu> 2018-01-04


zot_bib_web 2.0.4
-----------------------------------------

The program's configuration has been redesigned, and many more
options for displaying and sorting the interactive bibliography are available.


- Much simplified installation and setup.  Improved documentation.
  Settings files can now load multiple collections from different
  libraries.  Use user_collection(...) and group_collection(...)
  statements.  See settings_example.py for documentation.
  Previous settings file are imported (with exception of the next item).

- Documentation is now available at http://zot-bib-web.readthedocs.io .

- You can sort the bibliographic items not just
  by collection as in the Zotero library, but also by year or by
  publication type (e.g., journal article, conference paper), and
  arbitrary hierarchies thereof.   "Selected works" or "in
  review" papers can be extracted at the top.  Even if ordering chronologically,
  without showing thematic collections, the bibliography can be
  filtered according to area with the collection shortcuts at the top.

- Shortcuts at the top of the generated webpage can filter by years,
  collections, venues, publication type.

- Files (like PDF files) stored in the Zotero library can now be
  written out and linked to.

- Additional settings have become available.  See settings_example.py.

- Citations can be viewed in additional formats (see show_links
  configuration variable).  For example, we can show the full entry in
  APA or MLA format (or any other style available).

- Files attached to entries in Zotero can now be saved for upload to
  the server. (Requested by Selcuk Bilmis.)  HTML archives are also
  supported. Notes are supported in
  a similar way, but are not functional until some bugs in PyZotero or
  the Zotero API are fixed.

- Bibliographic entries can be numbered within their sections now.
  See the number_bib_items variable.

- Added several worked examples with output in the demo folder.
  Run make-demo.sh to re-generate them.

- Caching of Zotero retrieval makes repeated runs much faster.

- Tested with Pyzotero 1.2.11.
- Now compatible with Python 2.7 or Python 3.6.

Compatibility-breaking Changes:

- Catchallcollection is no longer available in program arguments or
the settings file.  Specify a "catch-all collection" via the &
modifier directly in the name of the respective collection,
e.g. "9999& Miscellaneous" (9999 if sorting at the end is desired).

- Output file is not longer a positional argument to the program.
  Use -o <outfile> instead.  '-' is permissible for stdout.

- Items in top-level collection are no longer excluded

- Limit has been removed (replaced with caching)



CSS  and HTML Changes:

- Search keywords are now prominently displayed when active

- full-bib-section and short-bib-section divs now contain their
respective section headings.

- Nested collections can be indented or otherwise styled using CSS
  (new div.collection tags)

- Several new class and id attributes to allow for more styling.

- Buttons (e.g., Wikipedia, EndNote, BIB) have standard capitalization
now.  Style using CSS.


Selected bug fixes:

- Fix problem with nested collections that did not show in the correct location
- Improved warning messages for double entries


zot_bib_web 1.2.2
-----------------------------------------
- Fix bug that prevented RIS downloads from showing
  (and potentially other elements included in show_links depending on capitalization)
- Improved instructions in settings example file
- Tested with Pyzotero 1.2.0 and Python 2.7


zot_bib_web 1.2.0
-----------------------------------------

- A "show to clipboard" button is shown for bib and wikipedia code (configure with show_copy_button entry)
- COINS information is now included.  This means Zotero's browser plugins will work with the generated bibliographies
- Wikipedia markup is included in the bibliography (configure with show_links setting)
- The shortened section (e.g., "Selected Works") now shows the journal or conference title  (to hide, use CSS style sheet)
- Visitors can now select text for copy/paste without also copying the meaningless buttons (configure with smart_selections setting)
- Style and js files necessary on the webserver are now in the "site" directory
- toplevelfilter may be None now, in which case the entire library is used.
- Many improvements to the generated HTML and supplied stylesheets
- Better IE11 support
- Tested with pyzotero 1.1.21

-- David Reitter <reitter@psu.edu> 2017-01-28


settings.example.py
## settings.example.py

## You must configure the authentication settings and one collection key.

#### AUTHENTICATION AND COLLECTIONS ######################################



group_collection(160464, collection='MGID93AS')

# user_collection(id='1366640', api_key='Y4oUU2c7P1nQWNPO3hvuEwzP')

# group_collection(id, api_key=None, collection=None, top_level=False)
# user_collection(id, api_key=None, collection=None, top_level=False)
#
#   Use group_collection for a group library, user_collection for a
#   (private) user library.
#   ID specifies the group or user ID.
#
#   You may find your user ID for the library_id setting under
#   "Settings -> Feeds/API":
#        https://www.zotero.org/settings/keys
#
#   You may find your library ID by selecting the group on the Zotero
#   website, and then choose "Group Settings".  The URL in your browser
#   window will then show you a six-digit number,
#   e.g.,  .../groups/110233/settings
#
#   api_key
#       The secret key provided by Zotero.
#       If you want to retrieve non-public data from Zotero, you'll need a
#       Zotero account (or group) at zotero.org.  Log into your account,
#       access the Settings page on the Zotero site and create an private
#       API key (under "Settings -> Feeds/API").  For the key, check
#       "Allow library access".  This key is used in the api_key setting.
#
#   collection
#       ID of the top-level collection to be included.
#       All sub-collections under this collection will be imported.
#       If not given (None), all available collections will be included.
#
#   top_level
#       If true, then the collection given be included as a level.
#       Otherwise (default, False), sub-collections and items will be
#       included directly.


#### COLLECTIONS TO BE RENDERED ##########################################

# In your library, you may create a collection, for example, "website".
# Within it, create titled sub-collections, like so:

# toRead
# thesis
# website
#    10 Selected Works
#    20 Journal Articles
#    30 Conference Proceedings
#    40 Theses

# Now, find the IDs of the top-level collection called "website".
# When you click on it on the Zotero website, your browser will show you
# an alphanumeric key in the URL, e.g., items/collectionKey/FCQM2AY6.
# The portion 'FCQM2AY6' is what you would use in 'collection' for
# the user_collection or group_collection directives.


# To cause zot_bib_web to format a sub-collection in special ways, you
# may add some statements here.  For these statements, supply either
# the name (title) or the key associated with a collection.

# exclude_collection (C)
# Remove sub-collection C.

# rename_collection (C, N)
# Rename collection C to N.
# This may be used to merge collections by giving them the same name.

# hidden_collection (C)
# Hide sub-collection C.  We still add a shortcut at the top
# to unhide its contents if they are available elsewhere.
# You may also use a '-' before the name of the collection in the library.

# short_collection (C)
# Short mode.  This sub-collection will be shown using titles,
# journal and years only, which can then be expanded.  Journal or
# conference titles can be kept short.  Specify the "journal abbr or
# "conference title" fields, or a short "note" if necessary.  You may
# want to copy bibliographic items from other parts of the bibliography
# into this sub-collection.
# You may also use a '*' before the name of the collection in the library.

short_collection("Selected Works")

# featured_collection (C)
# Feature this:  Extract this sub-collection and show at the beginning of
# the bibliography, regardless of whether the rest of the bibliography is
# sorted by, e.g., year, and ignores the collections otherwise. In the
# collection shown below, it prevents "in review" articles to show up as
# regular journal articles (which might give the impression you're
# taking credit for not-yet-reviewed/published material!)
# You may also use a '!' before the name of the collection in the library.

featured_collection("Selected Works")

# misc_collection (C)
# Show the items in this collection, but exclude those items that
# are already included in another regular collection.  A regular
# collection is one that is not hidden, not short, and not featured.
# This is usefull to add a "Miscellaneous" category at the end for
# additional items without duplicating anything.
# You may also use a '&' before the name of the collection in the library.

misc_collection("Miscellaneous")


#### SPECIAL OPTIONS: STYLE ##############################################
# Special settings - configure only if needed.

# The sort_criteria determine the structure of the bibliography.
# Allowable values: 'type' (category of pub., e.g., journal article),
# 'date' (full date)
# 'year' (year of publication)
# 'collection' (the subcollection the article is placed in).
# Collection works best at the beginning of the list.
# add - in front of the field name to sort in descending order
# (e.g., -date will show the newest entries first).

# sort_criteria = ['collection','-date']   # collection, then by date.
sort_criteria = ['type','-date']   # by type, then by date ("issued").
#sort_criteria = ['-year','type']   # Ordered by year, then type.



# bibliography title
titlestring = 'Bibliography'

# show section headings for the primary search criterion
show_top_section_headings = True

# bibliography style format (e.g., 'apa' or 'mla') - Any valid CSL style
# in the Zotero style repository
bib_style =  'apa'

#### SPECIAL OPTIONS: HTML AND FILE CONFIGURATION ########################
# Special settings - configure only if needed.


write_full_html_header = True
# False to not output HTML headers.
# In this case, expect a file in UTF-8 encoding.

stylesheet_url = "style.css"
# If set and write_full_html_header is True,
# link to this style sheet (a URL)

outputfile = 'zotero-bib.html'
# relative or absolute path name of output file
category_outputfile_prefix = 'zotero'
# relative or absolute path prefix

show_search_box = True
# show a Javascript/JQuery based search box to filter pubs by keyword.
# Must define jquery_path.

jquery_path = "site/jquery.min.js"
# path to jquery file on the server
# jquery_path = "../wp-includes/js/jquery/jquery.js"  # wordpress location

# Show categories at the top of the bibliography for quick filtering
# 'collection', 'type', 'year', 'venue', 'venue_short', or 'tags'
show_shortcuts = ['collection']
# To add specific rather than all available values for a field, use the
# shortcut function as follows.
# Note that for year, we support ranges.

# In the following, we tell the program
# to display despific entries for "year":

show_shortcuts += [shortcut('year', [2017,2016,2015,2014,2013,
                                     "2008-2012","2005-2008","-2004"])]

# Show the venues (conferences, journals) where we've published.
# the "shortcut" function offers some options:
# sortDir="asc" or "desc" instructs the program to sort the entries.
# (Default is automatic.)
# sortBy indicates that we'd like to sort by how many items there are
# in each category
#   (Default is to sort by name.  'count' is the only other value possible
#    at this time.)
# topN says to display only the 5 biggest categories.
# If there are several categories ranked 5, we're showing all of them.
# E.g.  topN=3 shows:  WIFN(8) FJJ(8) CLAM(5) CLEE(5)
# (Default is to show all for which bibliographic entries exist in the data.)

show_shortcuts += [shortcut('venue_short', sortDir='desc',
                            sortBy='count', topN=5)]

# To add arbitrary search terms:

show_shortcuts += [shortcut('keyword',
                            values=["model", "language", "entropy"])]
# define some search terms

number_bib_items = False
# show bibliographic items as numbered, ordered list

show_copy_button = True
# show clipbaord copy button.  Must define jquery_path.

clipboard_js_path = "site/clipboard.min.js"
# path to file on server

copy_button_path = "site/clippy.svg"
# path to file on server

# Attachments
file_outputdir = 'files/'
# Location to save attached files.

file_output_path = "files/"
# relative URL on server, corresponding to file_outputdir

# unconditionally show these items if they are available for the item
# (don't set to obtain defaults)
# show_links = ['abstract', 'url', 'bib', 'wikipedia',
#               'ris', 'cite.apa', 'file']
#
# URL       - the URL field for the item; button is displayed as
#             PDF/DOC/PS/LINK as detected.
# file      - each file associated with the item in the library.
#             Files are saved to file_outputdir and expected to appear
#             on the web server in file_output_path (which see)
#             Button is displayed as PDF/DOC/PS/LINK as detected.
# RIS or EndNote - download of the file containing the bibliographic data
# BIB       - bibliographic data for LaTeX
# Wikipedia - bibliographic data in Wikipedia format
# Abstract  - the abstract
# cite.APA  - Citation in APA format
# cite.MLA  - Citation in MLA format
# cite.<XXX> - Citation in <XXX> format (as supported by Zotero)



# omit_COinS = False
# True to omit COINS metadata; useful to save space, but not recommended.

# no_cache = True
# True means do not use cache;
# retrieve items from Zotero database every time (slow)


# smart_selections = True
# Prevent viewers from selecting "bib", "pdf" etc
# for easier copy/paste of bibliography
# (don't set to obtain default, which is True)

# Output in this language
language_code = 'en'

# Define labels for article types and their ordering
# types may occur in libraryCatalog or itemType
# use libraryCatalog to override it in special cases
# (e.g., archival Conference papers)
# sortkeyname_order['en']['type'] = [('journalArticle', 'Journal Articles'),
#                             ('archivalConferencePaper', 'Archival Conference Papers'),
#                             ('conferencePaper', 'Conference and Workshop Papers'),
#                             ('book','Books'),
#                             ('bookSection', 'Book Chapters'),
#                             ('edited-volume', "Edited Volumes"),
#                             ('thesis', 'Theses'),
#                             ('report', 'Tech Reports'),
#                             ('attachment', 'Document'),
#                             ('webpage', 'Web Site'),
#                             ('presentation', 'Talks')]

# Translations for links
# Provide additional languages like so:
# link_translations['de'] = {'abstract':'Abstrakt', 'pdf':'Volltext'}




#### SPECIAL OPTIONS: WORDPRESS ##############################
# Special settings - configure only if needed.
# These settings are used by push.py

push_wordpress(url='https://example.com/wp/xmlrpc.php', blogID=0,
               user='pubpushername', password='pass', postID=200)

infile = "zotero-bib.html"
